### ISSUE TITLE
Fixed lens light: whole-likelihood verdict on HST + Euclid × N × hardware (phase 5)

### ISSUE THREAD

**Issue by Jammy2211** [PyAutoLabs/autolens_profiling#259]
## Overview

Phase 5 — the last — of the `fixed-lens-light-profiling` epic, and the assessment the
human actually asked for: for each hardware type, what does the **whole likelihood
function** cost on real survey data with the lens light fixed, and which configuration
should production use. Phases 0–4 measured one axis at a time (kernels, the library row,
hardware, model quality, source-pixel scaling). Nothing in the epic so far answers "which
one do we ship", and nothing in it has ever touched Euclid.

The grid is the library path (`AnalysisImaging.log_likelihood_function` under `jax.jit`)
on **HST (0.05") and Euclid (0.1")** at **500 / 1250 / 2500** source pixels, across
**A100 fp64 / RTX 2060 fp64 / RTX 2060 mixed precision / JAX-CPU (8 threads)**, for the
solvers phase 4 certified as affordable. The deliverable is
`results/notes/fixed_lens_light_verdict_2026_09.md`: the table, and then a written
production configuration per hardware — solver, pass budget, precision and the source
pixel count that hardware can afford — with the evidence cost of every shortcut stated in
**nats**, not only in milliseconds.

Dense inversion only; the sparse operator and JWST are out of scope for the whole epic by
the human's own clarification.

## Plan

- Extend the phase-1 cell `scripts/imaging/likelihood_breakdown/fixed_light_library.py`
  with a `--dataset {hst,euclid}` switch (HST default byte-identical), `--pass-budget auto`
  (resolve the production budget per mesh and record the measured certifying budget beside
  it) and `--routes a,b,c,d` (skip `d0`/`e` to halve leg time).
- Time the **production** certified row at phase 3's safe budgets — **11** (rectangular) /
  **7** (Delaunay) — not at whatever budget happens to certify, and record the per-config
  certifying budget from a kernel budget sweep over 1..12 beside it. Phase 4 showed the
  certifying budget does not scale with N (rect 5→10, Delaunay 1→2 over 484→3969 pixels),
  so the sweep is cheap insurance and the only Euclid evidence that will exist.
- **Euclid has no pins.** Its log determinants, evidences and equivalence comparisons are
  **recorded, not asserted** — the first Euclid reference for this family. HST at
  N=1250/2500 is off the fiducial mesh and records likewise.
- 12 A100 legs (rect + Delaunay × 2 datasets × 3 N) submitted early to `gpu-2`; 18 local
  legs (Delaunay only — phase 4 answered rect on consumer hardware) run one at a time on a
  quiet machine: RTX 2060 fp64, RTX 2060 mixed precision, JAX-CPU at `NPROC=8` / BLAS 1.
- Write `results/notes/fixed_lens_light_verdict_2026_09.md`: the 2 × 3 × hardware table per
  solver, the per-hardware production configuration (solver, budget + fallback, precision,
  affordable N, ms/call, nats against S0 today), the consumer-GPU memory ceiling, what the
  programme did **not** measure, and the follow-ups the verdict implies — filed as
  `ideas.md` bullets in PyAutoMind, not as new prompts.
- Stacked PR against `feature/fixed-light-scaling` (a stack of six). Merge stays human.

<details>
<summary>Detailed implementation plan</summary>

### Work Classification
Workspace (`autolens_profiling` — scripts, results, notes; no library source).

### Affected Repositories
- autolens_profiling (primary)

### Branch Survey
| Repository | Current Branch | Dirty? |
|-----------|---------------|--------|
| ./autolens_profiling | main | clean |

**Suggested branch:** `feature/fixed-light-verdict`
**Worktree root:** `~/Code/PyAutoLabs-wt/fixed-light-verdict/`
**Base:** `feature/fixed-light-scaling` (phase 4, PR #258) — a stack of six.

### Implementation Steps

1. `scripts/imaging/likelihood_breakdown/fixed_light_library.py`
   - `--dataset {hst,euclid}` (default `hst`): selects the `INSTRUMENTS` preset (pixel
     scale 0.05 → 0.1, dataset dir, auto-simulate if missing). Everything else — mask
     radius 3.5", the `[4,2,2]`/`[0.3,0.6]` light-profile over-sampling, the 60×1 MGE, the
     mass/shear priors, the regularization — is held byte-identical across datasets, which
     is what makes the two columns comparable. The HST defaults are unchanged.
   - `--pass-budget auto`: run the kernel certification sweep first (budgets 1..12 on the
     config's own S3 system, `active_set_masked_jax` at the leg's `tau_rel`), record the
     smallest certifying budget, and time route `d` at the **production** safe budget
     (phase 3: 11 rectangular / 7 Delaunay, overridable with `--safe-budget`).
   - `--routes a,b,c,d`: route filter; `d0`/`e` are phase-1 diagnostics and are skipped
     here so a leg is four library rows, not six.
   - Pins: `PIN_IS_FIDUCIAL` is false at every N in this grid and on Euclid at any N, so
     every leg records rather than asserts; a `reference_recorded` block writes the
     log determinants and evidences as the first Euclid reference.
2. `scripts/misc/test/test_fixed_light_verdict.py` — `--pass-budget auto` picks the
   smallest certifying budget from a hand-made pass table; `--dataset` defaults to HST and
   the HST path is unchanged; the route filter admits exactly `a,b,c,d`; the safe budgets
   are 11/7.
3. `hpc/batch_gpu/submit_breakdown_imaging_fixed_light_verdict_{pixelization,delaunay}_a100_{hst,euclid}_fp64`
   (4 arrays × 3 arms = 12 legs) + `hpc/batch_gpu/submit_fixed_light_verdict.sh`, mirroring
   `submit_fixed_light_library.sh`. RAL worktree
   `/mnt/ral/jnightin/autolens_profiling_wt/fixed-light-verdict`, submitted only once the
   checkout has fully completed (phase 4 lost ten arms to a premature `sbatch`).
4. Local legs: 18, one at a time, fresh JAX compilation cache each, Delaunay only.
   CPU N=2500 capped at 30 min (a kill is a recorded result, not a gap).
5. Harvest → `results/notes/fixed_lens_light_verdict_2026_09.md` → `ideas.md` bullets →
   README regen → stacked PR.

### Key Files
- `scripts/imaging/likelihood_breakdown/fixed_light_library.py` — the phase-1 cell, extended
- `scripts/misc/likelihood_breakdown/library_solver_injection.py` — the certified-solver injection
- `scripts/misc/likelihood_breakdown/active_set_steps.py` — the kernels and the S3 builder
- `results/notes/fixed_lens_light_verdict_2026_09.md` — the programme deliverable
- `instruments/imaging.py` — the Euclid preset (0.1", mask 3.5")

### Gates
`ruff check` + `ruff format --check`, the cell's `AUTOLENS_PROFILING_SMOKE` import
short-circuit, `build_readme.py --check`, `check_submits.py --check`, the full
`scripts/misc/test` suite (phase-4 count 348 + new).

</details>

## Original Prompt

<details>
<summary>Click to expand starting prompt</summary>

# Fixed lens light — whole-likelihood assessment on HST and Euclid at 500 / 1250 / 2500 source pixels

Type: research
Target: autolens_profiling
Repos:
- autolens_profiling
Themes:
- profiling
- pixelization
- hpc-gpu
Difficulty: large
Autonomy: supervised
Priority: high
Status: formalised
Consequence: judge
Witness: results/notes 2 datasets x 3 source-pixel counts x 3 hardware table per solver (S3-certified / S3-unconstrained / S0-today), and a stated production configuration per hardware
Review-minutes: 25
Unattended: needs-slicing
Epic: fixed-lens-light-profiling
Phase: 5
Filed: 2026-09-13

Phase 5 of the `fixed-lens-light-profiling` epic — the programme's verdict. Gated on
phase 4 — "one step at a time".

## Why

Phases 1-4 measure solvers, hardware and scaling one axis at a time. Phase 5 is the
assessment the human actually asked for: for each hardware type, what does the whole
likelihood function cost on real survey data, and which configuration should production
use. Nothing else in the epic answers "which one do we ship".

## What to do

Whole-likelihood assessment on the **library path** (`FitImaging.figure_of_merit` call, not
the kernel harness) across the grid:

- Datasets: HST (0.05", `autolens_profiling/dataset/imaging/hst`) and Euclid (0.1",
  `.../euclid`). JWST is dropped by human decision.
- Source pixels: 500, 1250, 2500.
- Hardware: A100 (RAL gpu-2, fp64), CPU, RTX 2060 via the `PyAutoGPU` venv (fp64 and
  fp32/mp).
- Solvers: S3 with certified positivity, S3 unconstrained, and S0 as it stands today.

Dense inversion only — the sparse operator is out of scope for this epic (the PyAutoArray
weight-map bug, `draft/bug/autoarray/sparse_inversion_ignores_profile_subtracted_image.md`),
even though the human's original text named it as a variant.

The note is the programme's deliverable: the full table, and then a written verdict naming
the production configuration per hardware — solver, pass budget, precision, and the source
pixel count each hardware can actually afford — with the evidence cost of every shortcut
stated in nats, not just in ms.

Cell `@autolens_profiling/scripts/imaging/likelihood_breakdown/fixed_light.py`
(`--library-row`). The A100 submit → wait → harvest step is a human resume point, not a park.
## Original request (2026-09-13, verbatim — the epic request)

"Ok these are the takss to now queue up: 1) Perform profiling of positive-negative solver for completeness albeit S3 certified
active-set seems like the lead forward; 2) Perform the same profiling on CPU (using CPU appropriate methods) and on m laptop GPU as
its important now we are converging on a solutiopn we also optimizze for consumer GPUs; 3) Confirm that S3 certified active-set is
ver fast on solutions which give low likelihoods, as maybe the fast run times here are a result of the model being good and giving
high likelihoods or an easier to solve solution; 4) Provide profiling information over number of source pixels with the new approach
on difernet hardware 5) For all hardware types and likelihood variants (e.g. sparse operator) give an assessment of the overall
likelihood function on JWST (0.03" pixel scale), HST and Euclid data for 500, 1250 and 2500 source pixels. For each take one step
at a time, and try do it all in --auto."

Clarification (same day): "For positive-negative I want you to use a positive negative solve (e..g. np.linalg.solve) but DO NOT include the
MGE in the matrices or solution, so it sould use normal light porfiles which subtract the MGE beforehand. PyAutoGPU is avirtual
enviroment which already exists on this laptop for running GPU JAX so use that. Rest of plan sonds good. Only use HST and Euclid
for now, drop JWST. Skip the sparse operator for now."

</details>

**Comment by Jammy2211** [comment]
## Shipped — 2026-09-13

### PRs
- https://github.com/PyAutoLabs/autolens_profiling/pull/260 (stacked on #258 — merge order #250 → #252 → #254 → #256 → #258 → #260)

### Summary

Phase 5 — the last — of the `fixed-lens-light-profiling` epic. The whole library likelihood
call on **HST and Euclid** at 500 / 1250 / 2500 source pixels, four routes per leg, with a
production configuration stated per hardware and every shortcut's cost in **nats**. Note:
`results/notes/fixed_lens_light_verdict_2026_09.md`.

**24 local legs, 24 JSONs, zero non-zero exits, zero timeouts, 45 minutes.** The twelve A100
legs of the grid were **not run** — the RAL jump host refused publickey for the whole session
and the direct route times out. The four arrays are written, validated (`bash -n`,
`check_submits.py`) and committed; the resume is one command on the RAL login node, in a
worktree of this branch, **after the checkout completes**:

```
hpc/batch_gpu/submit_fixed_light_verdict.sh --node euclid-ral-gpu-2
```

Every A100 number in the note is carried from phases 1 and 4 and labelled as such.

### Findings

1. **Euclid is where positivity earns its keep, and HST never showed it.** Dropping positivity
   costs **+3.5 → +24.3 → +109.1 nats** as N goes 500 → 2500 on Euclid (Delaunay) against a
   *flat* +8.4 → +6.4 on HST, with 129 negative source pixels against HST's 6. The cause is the
   constraint ratio — at N=2500 Euclid has 1.5 image pixels per source pixel where HST has 6.1
   — and the trend with N **reverses** between the datasets. Every earlier phase of this epic
   measured that shortcut on HST alone.
2. **The certified active set is the production solver everywhere**, 1.20–1.59× over what the
   library runs today on HST and 1.43–2.28× on Euclid, where the lead *grows* with N. It
   returns the library's own answer (every equivalence pin PASS, max 5.6e-10 relative), so its
   cost in nats is **zero**.
3. **The pass budget is dataset-dependent.** Rectangular certifies at 6 / 9 / 11 passes on
   Euclid against 5 / 7 / 10 on HST — phase 3's safe budget of **11, exhausted exactly** at
   N=2500, from a single model. Delaunay keeps its margin on both (Euclid 2 / 4 / 4, HST
   1 / 2 / 1, budget 7) and stays the production mesh. Phase 4's "the budget does not scale
   with N" holds; it does move with the dataset.
4. **Nothing is memory- or precision-bound**: 2.70 GB peak of 6 GB, and mixed precision buys
   4–9 % for ≤ 1.3e-4 nats with an identical certifying budget to fp64 on Delaunay.

**Affordable N, in the recommended configuration:** 4000 HST (A100, carried from phase 4) /
1500 HST and 2500 Euclid (RTX 2060) / 1000 HST and 1250 Euclid (8 CPU threads).

### Session notes

- Two independent ties back to phase 4 rather than one: the HST N=2500 S3 PDIP row (2241 ms
  here against 2225 ms there) and the rectangular route-c evidence (+318.00 / +347.11 nats
  here against +318.0 / +347.1 measured on the kernel path).
- The six rectangular RTX legs were **not** in the plan — the plan put rectangular on the A100.
  They were added when the A100 could not be reached, and they are what found finding 3.
- `--pass-budget auto` deliberately resolves to the **production** budget (phase 3's
  zero-fallback 11 / 7), not to whatever certifies at that configuration; the certifying budget
  is measured by the sweep and recorded beside it, which is the only Euclid evidence that
  exists on that question.
- Five follow-ups are filed as `ideas.md` bullets, not prompts: the library implementation of
  the certified solver (a `Settings` option with the PDIP fallback and a cond-free batched
  path), the matched-injection witness before any unconstrained use, the Delaunay assembly as
  the next GPU lever, re-deriving the rectangular budget on Euclid over a graded draw set, and
  running the twelve A100 legs.
- **With this merged, the `fixed-lens-light-profiling` epic is complete.** The epic ledger's
  `status:` line is deliberately left as-is until the human merges the stack.

**Comment by Jammy2211** [comment]
Shipped — PR #260 merged to `main`. **Phase 5, the last of the `fixed-lens-light-profiling` epic — the epic is COMPLETE.**

24 local legs, all rc=0, 45 minutes total: Delaunay × {HST, Euclid} × {500, 1250, 2500} on RTX 2060 fp64, RTX 2060 mp and JAX-CPU, plus 6 rectangular RTX fp64 legs.

**Euclid is where positivity earns its keep.** Dropping it costs **+3.5 → +24.3 → +109.1 nats** as N goes 500 → 2500 on Euclid, against a flat **+8.4 → +6.4** on HST (129 negative source pixels vs 6) — because at N=2500 Euclid has 1.5 image pixels per source pixel where HST has 6.1. The trend with N reverses between the datasets, and every earlier phase measured this shortcut on HST alone.

The certified active set is the production solver on every hardware and both datasets (1.20–1.59× on HST, 1.43–2.28× on Euclid where the lead grows with N) and returns the library's own answer to 5.6e-10, so its cost in nats is **zero**. The pass budget is dataset-dependent: rectangular certifies at 6/9/11 on Euclid against 5/7/10 on HST, exhausting phase 3's safe budget of 11 exactly at N=2500 from a single model. Delaunay keeps its margin and stays the production mesh.

**Open gap:** the 12 A100 legs were not run — the RAL jump host refused publickey all session. The four arrays are written, validated and committed; resume is one command: `hpc/batch_gpu/submit_fixed_light_verdict.sh --node euclid-ral-gpu-2`.

Note: `results/notes/fixed_lens_light_verdict_2026_09.md`.

Two follow-on research programmes are being filed from this verdict: the same programme on the **numba CPU path**, and an optimisation campaign against the **non-solver residue** of the HST GPU breakdown (≈21 ms of the 25 ms certified Delaunay call).
